Add Ironic enrolment

w-miller opened the pull request:
Create bare metal nodes in Ironic for each node specified, and also bare metal ports in Ironic for each NIC.

Commit:
Extract IPMI configuration out of role

This is required for Ironic enrolment as well as Virtual BMC
configuration.

Commit:
Consolidate interface/link naming

Create filter plugins to perform physnet/node to veth link and bridge
naming. This reduces the logic duplication present.

Commit:
Add additional network naming filter plugins

These will be useful for determining details needed for Ironic ports.

Commit:
Ensure nodes are always sorted alphabetically

Commit:
Add Ironic enrolment Ansible role

markgoddard commented on ansible/roles/virtualbmc-domain/defaults/main.yml:
```diff
@@ -1,13 +1,12 @@
 ---
 # The address on which VBMC will listen for IPMI traffic for this domain.
-vbmc_ipmi_listen_address: 0.0.0.0
+vbmc_listen_address:
```
vbmc_ipmi_address

w-miller replied:
Good spot.

markgoddard commented on ansible/deploy.yml:
```diff
@@ -77,10 +79,37 @@
         loop_var: domain
         index_var: port_offset
+
+- hosts: localhost
```
New playbook for this?

w-miller replied:
Sure.

w-miller replied:
Actually, it looks like we've got more faffing for Virtual BMC above that's currently in deploy.yml than we have here for Ironic enrolment: should I move that into a separate file too? And incidentally, would it be best to make these separate playbooks with the hosts declaration inside them, or separate task files with the hosts declaration still in deploy.yml?

markgoddard replied:
The main thing I'd like is to be able to run each stage separately. That would require separate playbooks, or alternatively tags. I might also want to skip enrolment entirely, in favour of autodiscovery with ironic inspector.

w-miller replied:
That makes sense. I've split things out a bit and this is the architecture I've got now - the blue rectangles are playbooks which would be callable individually. Is this as granular as you'd like?
![tenks ansible structure 1](https://user-images.githubusercontent.com/14925373/45108687-35df2680-b135-11e8-8017-e3c0d98ef3c4.png)

markgoddard replied:
Awesome diagram! That should definitely go in the blog post and docs.

That's definitely closer. I'd still like to be able to skip ironic enrolment. I guess we could do that with a flag, or if you move ironic_enrolment.yml up to deploy.yml that would work.

markgoddard replied:
Does `enrol_nodes.yml` fit better with the playbook naming scheme?

w-miller replied:
Ah yes, true. Probably fits best as a third 'stage' to deploy.yml, so I'll move it up a level.

w-miller replied:
And yes, `enrol_nodes.yml` would be a good name for it on the second level down.

w-miller replied:
Updated diagram:
![tenks ansible structure 4](https://user-images.githubusercontent.com/14925373/45148059-50111700-b1be-11e8-81aa-31149fade925.png)

markgoddard commented on ansible/deploy.yml:
```diff
@@ -77,10 +79,37 @@
+        ironic_virtualenv_path: "{{ virtualenv_path }}"
+        ironic_python_upper_constraints_url: >-
+          {{ python_upper_constraints_url }}
+      loop: "{{ allocations.result.iteritems() | list }}"
```
How about using the dict query?

`loop: "{{ query('dict', allocations.result }}"`

Then you can reference `alloc.key` and `alloc.value`.

w-miller replied:
Ah, I didn't know about that one. Thanks.

markgoddard commented on ansible/filter_plugins/tenks.py:
```diff
@@ -0,0 +1,183 @@
+    custom Python modules, so we can't have a Libvirt filters file that imports
+    a network filters file; for the same reason, we can't have a shared utils
+    file either. This is why all Tenks filters are lumped together in this
+    file.
```
One way around this is to use the tenks python module, and import that. At the moment it's not being used though to might be simpler to keep it like this.

w-miller replied:
Good to know, thanks.

markgoddard commented on ansible/filter_plugins/tenks.py:
```diff
@@ -0,0 +1,183 @@
+    :param size: The size string to parse.
+    :returns: The number of bytes represented by `size`, as an integer.
+    """
+    UNITS = {"": 1, "K": 10**3, "M": 10**6, "G": 10**9, "T": 10**12}
```
Should these be 2^10 rather than 10^3?

w-miller replied:
Actually, since we're passing our capacities to libvirt-vm, which passes its capacities to virsh, we should probably be following [virsh's convention](https://linux.die.net/man/1/virsh). This specifies K,M etc to refer to kilo/megabytes (10^x), and KB,MB etc to refer to kibi/mebibytes (2^y) - I haven't made that distinction here.

markgoddard commented on ansible/filter_plugins/tenks.py:
```diff
@@ -0,0 +1,183 @@
+    except KeyError:
+        msg = ("The size string '%s' contains an invalid unit '%s'. Valid "
+               "units are: %s." % (size, unit, ", ".join(UNITS.keys())))
+        raise AnsibleFilterError(to_text(msg))
```
For future reference, https://pint.readthedocs.io/en/latest/ is useful. It would add a python dependency to tenks however.

w-miller replied:
Looks like a useful library! Probably would be a bit overkill for a single function here though :)

Commit:
Split deploy.yml

Split deploy.yml into host- and node-specific steps.

Conflicts:
	ansible/deploy.yml

markgoddard commented on ansible/roles/ironic-enrolment/tasks/main.yml:
```diff
@@ -0,0 +1,18 @@
+---
+- name: Ensure Python requirements are installed
+  pip:
+    requirements: "{{ '/'.join([role_path, 'files', 'requirements.txt']) }}"
```
did you try just 'requirements.txt' here?

w-miller replied:
I did, yes. It seems like while Ansible magically checks the role's *files* directory for the source field in `copy`/`template` actions, it doesn't work in the `pip` module.

Commit:
Fix PEP8 issues

markgoddard commented on ansible/roles/ironic-enrolment/tasks/node.yml:
```diff
@@ -0,0 +1,90 @@
+      # volume, make this configurable?
+      disk_size: >-
+        {{ (node.volumes.0.capacity | default('1')) | size_string_to_gb }}
+      cpus: "{{ node.vcpus }}"
```
We'll soon want to avoid registering properties, and use the resource class instead.

w-miller replied:
Okay. Should I leave the properties in here for now, and add flavor creation and resource class linking as well?

markgoddard replied:
Actually, let's remove them now. They shouldn't be necessary, and can cause issues in environments with VMs and bare metal.

w-miller replied:
Sure.

w-miller replied:
I'll probably do the flavors/resource classes in a separate pull request too, to avoid cluttering this one any more.

markgoddard commented on ansible/roles/ironic-enrolment/tasks/node.yml:
```diff
@@ -0,0 +1,90 @@
+    # This module requires the openstacksdk package, which is installed within
+    # our virtualenv.
+    ansible_python_interpreter: >-
+      {{ '/'.join([ironic_virtualenv_path, 'bin', 'python']) }}
```
Glad to see this works now. Up to ansible 2.4 ansible_python_interpreter would not work with jinja.

markgoddard reviewed (changes requested):
Looks like a good first pass.

Commit:
Fix IPMI address typo

Commit:
Use dict query instead of iteritems

Commit:
Allow binary and decimal capacities in size string

Commit:
Split deployment into stages

Commit:
Fix PEP8 and ansible-lint warnings

Commit:
Merge remote-tracking branch 'origin/master' into ironic-enrolment

Conflicts:
	ansible/deploy.yml

w-miller commented:
Updated and ready for re-review.

Commit:
Move node enrolment up a level

Call node enrolment from the top level deploy.yml playbook, to make it
easier to skip enrolment if necessary. Rename to fit the scheme of
second-level playbooks.

Also, fix variable scoping issues by giving a different name to the
temporary variable used when calling the scheduling module. This ensures
that all future tasks correctly reference the allocations from the file.
